src/batched/KokkosBatched_Gemm_Decl.hpp

```diff
@@ -1,114 +1,229 @@
 
-}
+/* BEGIN host-level routines */
+
+// Forward declare BatchedSerialGemm
+template <typename ArgTransA, typename ArgTransB, typename ArgMode,
+          typename ArgBatchLayout, typename ArgResultsPerThread>
+struct BatchedSerialGemm;
+
+/// \brief Non-blocking solve of general matrix multiply on a batch of
+/// matrices.
+///        C = alpha * op(A) * op(B) + beta * C
+///
+/// \tparam ArgTransA      Specifies what op does to A:
+///                        Trans::NoTranspose for non-transpose
+///                        Trans::Transpose for transpose
+///                        Trans::ConjTranspose for conjugate transpose
+/// \tparam ArgTransB      Specifies what op does to B:
+///                        Trans::NoTranspose   for non-transpose
+///                        Trans::Transpose     for transpose
+///                        Trans::ConjTranspose for conjugate transpose
+/// \tparam ArgBatchLayout Specifies where the batch dimension is allocated in
+///                        AViewType, BViewType, and CViewType:
+///                        BatchLayout::Left  Batch dimension is the leftmost
+///                        dimension BatchLayout::Right Batch dimension is the
+///                        rightmost dimension
+/// \tparam ScalarType     Specifies the scalar type of alpha and beta
+/// \tparam AViewType      Input matrix, as a 3-rank Kokkos::View
+/// \tparam BViewType      Input matrix, as a 3-rank Kokkos::View
+/// \tparam CViewType      Input(RHS)/Output(LHS) matrix, as a 3-rank
+/// Kokkos::View
+///
+/// \param alpha [in]      Input coefficient used for multiplication with A
+/// \param A [in]          Input matrix, as a 3-rank Kokkos::View
+///                        If ArgBatchLayout == "BatchLayout::Right", matrix A
+///                        is MxKxB If ArgBatchLayout == "BatchLayout::Left",
+///                        matrix A is BxMxK
+/// \param B [in]          Input matrix, as a 3-rank Kokkos::View
+///                        If ArgBatchLayout == "BatchLayout::Right", matrix B
+///                        is KxNxB
+/// \param beta [in]       Input coefficient used for multiplication with C
+///                        If ArgBatchLayout == "BatchLayout::Left",  matrix A
+///                        is BxKxN
+/// \param C [in/out]      Input/Output matrix, as a 3-rank Kokkos::View
+///                        If ArgBatchLayout == "BatchLayout::Right", matrix C
+///                        is MxNxB If ArgBatchLayout == "BatchLayout::Left",
+///                        matrix C is BxMxN
+/// \return 0 upon success, non-zero otherwise
+template <typename ArgTransA, typename ArgTransB, typename ArgBatchLayout>
+struct BatchedGemm {
+  template <typename ScalarType, typename AViewType, typename BViewType,
+            typename CViewType>
+  static int invoke(const ScalarType alpha, const AViewType &A,
+                    const BViewType &B, const ScalarType beta,
+                    const CViewType &C) {
```

Comment:
@e10harvey Currently, ```BatchedGemm<...>::invoke()``` only calls ```BatchedSerialGemm<...>::invoke()```. Should we also have a similar interface for batched gemm using shared memory?

Reply:
The idea is to have a single `BatchedGemm` interface that will invoke the correct underlying interface for optimal performance. We will likely invoke a `BatchedTeamVectorGemm` via `BatchedGemm` in the future.

Reply:
You mean we will have a selective interface of ```BatchedGemm``` for either ```BatchedSerialGemm``` or ```BatchedTeamVectorGemm```, right?

Reply:
CC: @srajama1 

@vqd8a Yes. Except, the current plan is to have two internal interfaces `BatchedSerialGemmInternal` and `BatchedTeamVectorGemmInternal` that are invoked by the selective interface, `BatchedGemm`.